gio/src/subclass/list_model.rs

```diff
@@ -0,0 +1,83 @@
+use crate::auto::ListModel;
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use glib::Cast;
+use glib::ObjectExt;
+
+pub trait ListModelImpl: ObjectImpl {
+    fn get_item_type(&self, list_model: &Self::Type) -> glib::Type;
+    fn get_n_items(&self, list_model: &Self::Type) -> u32;
+    fn get_item(&self, list_model: &Self::Type, position: u32) -> Option<glib::Object>;
+}
+
+unsafe impl<T: ListModelImpl + glib::ObjectType> IsImplementable<T> for ListModel {
+    unsafe extern "C" fn interface_init(
+        iface: glib::ffi::gpointer,
+        _iface_data: glib::ffi::gpointer,
+    ) {
+        let list_model_iface = &mut *(iface as *mut ffi::GListModelInterface);
+
+        list_model_iface.get_item_type = Some(list_model_get_item_type::<T>);
+        list_model_iface.get_n_items = Some(list_model_get_n_items::<T>);
+        list_model_iface.get_item = Some(list_model_get_item::<T>);
+    }
+}
+
+unsafe extern "C" fn list_model_get_item_type<T: ObjectExt + ListModelImpl>(
+    list_model: *mut ffi::GListModel,
+) -> glib::ffi::GType {
+    let instance = &*(list_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let type_ = imp
+        .get_item_type(from_glib_borrow::<_, ListModel>(list_model).unsafe_cast_ref())
+        .to_glib();
+    // Store the type so we can enforce that it doesn't change.
+    let key = glib::Quark::from_string("type");
+
+    match imp.get_qdata(key) {
+        Some(old_type) => {
+            if type_ != *old_type {
+                panic!("ListModel's get_item_type cannot be changed");
+            }
+        }
+        None => {
+            imp.set_qdata(key, type_);
+        }
+    }
+    type_
+}
+
+unsafe extern "C" fn list_model_get_n_items<T: ListModelImpl>(
+    list_model: *mut ffi::GListModel,
+) -> u32 {
+    let instance = &*(list_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    imp.get_n_items(from_glib_borrow::<_, ListModel>(list_model).unsafe_cast_ref())
+}
+
+unsafe extern "C" fn list_model_get_item<T: ObjectExt + ListModelImpl>(
+    list_model: *mut ffi::GListModel,
+    position: u32,
+) -> *mut glib::gobject_ffi::GObject {
+    let instance = &*(list_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let type_ = imp
+        .get_qdata::<glib::Type>(glib::Quark::from_string("type"))
+        .expect("ListModel should define the item-type");
```

Comment:
Maybe `get_item_type()` was not called before this, in which case you probably want to initialize it here

Reply:
I added a FIXME comment for that into the code

Reply:
I just ended up using get_item_type instead of reading it from the qdata directly